### PR TITLE
tests: Add fuzzing harnesses for CAutoFile, CBufferedFile, LoadExternalBlockFile and other FILE* consumers

### PR DESCRIPTION
**Pull request by practicalswift**
Add fuzzing harnesses for `CAutoFile`, `CBufferedFile`, `LoadExternalBlockFile` and other `FILE*` consumers:
* Add `FuzzedFileProvider` which provides a `FILE*` interface to `FuzzedDataProvider` using `fopencookie`
* Add `FuzzedAutoFileProvider` which provides a `CAutoFile` interface to `FuzzedDataProvider`
* Add serialization/deserialization fuzzing helpers `WriteToStream(…)`/`ReadFromStream(…)`
* Add fuzzing harness for `CAutoFile` (`streams.h`)
* Add fuzzing harness for `CBufferedFile` (`streams.h`)
* Add fuzzing harness for `LoadExternalBlockFile(...)` (`validation.h`)
* Add fuzzing harness for `CBlockPolicyEstimator::Read` and `CBlockPolicyEstimator::Write` (`policy/fees.h`)

See [`doc/fuzzing.md`](https://github.com/bitcoin/bitcoin/blob/master/doc/fuzzing.md) for information on how to fuzz Bitcoin Core. Don't forget to contribute any coverage increasing inputs you find to the [Bitcoin Core fuzzing corpus repo](https://github.com/bitcoin-core/qa-assets).

Happy fuzzing :)

### CODE REVIEW

**Review by Crypt-iQ — COMMENTED** [review]
utACK 27bef8c

File: src/test/fuzz/util.h [hunk 1]
```diff
@@ -228,4 +229,121 @@ NODISCARD inline std::vector<uint8_t> ConsumeFixedLengthByteVector(FuzzedDataPro
...
+        if (buf == nullptr || size == 0 || fuzzed_file->m_fuzzed_data_provider.ConsumeBool()) {
+            return fuzzed_file->m_fuzzed_data_provider.ConsumeBool() ? 0 : -1;
+        }
+        std::memset(buf, 0, size);
+        const std::vector<uint8_t> random_bytes = fuzzed_file->m_fuzzed_data_provider.ConsumeBytes<uint8_t>(size);
+        if (random_bytes.empty()) {
+            return 0;
+        }
+        std::memcpy(buf, random_bytes.data(), random_bytes.size());
+        const ssize_t n = fuzzed_file->m_fuzzed_data_provider.ConsumeBool() ? size : random_bytes.size();
```
[the remark this answers is not in the document]
Comment (MarcoFalke) [reply src/test/fuzz/util.h:300]: If the m_fuzzed_data_provider is empty it should give out less bytes, so `size` would be larger. `ConsumeBool` would return false in that case and this condition would evaluate to `size`.

So it seems `n` can simply be replaced by `size`

[the remark this answers is not in the document]
Comment (Crypt-iQ) [reply src/test/fuzz/util.h:300]: Oh I didn't think about that edge case. Don't you mean the condition should evaluate to `random_bytes.size()` if `ConsumeBool()` is false. So `n` replaced by `random_bytes.size()` instead?

[the remark this answers is not in the document]
Comment (MarcoFalke) [reply src/test/fuzz/util.h:300]: Ah right off-by-one :man_facepalming:

**Review by Crypt-iQ — COMMENTED** [review]
Combined fuzzing coverage for only the fuzzers in *this* branch: https://crypt-iq.github.io/pr19143_cov/src/

Each fuzzer was run for about 24 hours. I used a variation of libFuzzer and AFL. 
There seems to be some missing coverage in `CBlockPolicyEstimator::Write` & `CBlockPolicyEstimator::Read`. That could probably be fixed by better seeds. There's also missing coverage in `LoadExternalBlockFile`, and this can also be fixed by better seeds. I don't know the format of the external block files (I copy pasted the magic bytes from `MessageStart()` as seeds but no luck). This was the only slow fuzzing harness with ~5 execs/s.

File: src/test/fuzz/buffered_file.cpp [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+            switch (fuzzed_data_provider.ConsumeIntegralInRange<int>(0, 4)) {
+            case 0: {
+                std::array<uint8_t, 4096> arr{};
+                try {
+                    opt_buffered_file->read((char*)arr.data(), fuzzed_data_provider.ConsumeIntegralInRange<size_t>(0, 4096));
+                } catch (const std::ios_base::failure&) {
+                }
+                break;
+            }
+            case 1: {
```
[the remark this answers is not in the document]
Comment (practicalswift) [reply src/test/fuzz/buffered_file.cpp:43]: Good catch! Now fuzzing also `SetPos` :)

**Comment by practicalswift** [comment]
> Combined fuzzing coverage for only the fuzzers in _this_ branch: https://crypt-iq.github.io/pr19143_cov/src/
> 
> Each fuzzer was run for about 24 hours. I used a variation of libFuzzer and AFL.

Wow, that's some thorough testing! Thanks a lot!

> There seems to be some missing coverage in `CBlockPolicyEstimator::Write` & `CBlockPolicyEstimator::Read`. That could probably be fixed by better seeds.

I think the perceived lack of coverage is simply due to `src/test/fuzz/policy_estimator` being very slow (`CBlockPolicyEstimator` creation is costly): I believe that given enough time the fuzzer will span the full CFG :)

To speed things up I've now added also a dedicated fuzzer for `CBlockPolicyEstimator::{Write,Read}` which re-uses the `CBlockPolicyEstimator` object across runs (without sacrificing substantial coverage stability AFAICT): see `src/test/fuzz/policy_estimator_io.cpp`.

This dedicated fuzzer is running at >1000 exec/s compared to <50 exec/s for the non-dedicated one.

> There's also missing coverage in `LoadExternalBlockFile`, and this can also be fixed by better seeds. I don't know the format of the external block files (I copy pasted the magic bytes from `MessageStart()` as seeds but no luck). This was the only slow fuzzing harness with ~5 execs/s.

libFuzzer from LLVM 10 is able to reach past the magic bytes check within a few seconds. Perhaps you were using an old version of libFuzzer when running the `load_external_block_file` fuzzer?

File: src/test/fuzz/util.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (practicalswift) [reply src/test/fuzz/util.h:336]: Now simply using `random_bytes.size()`. I think I was trying to be too clever :)  Please re-review :)

**Review by Crypt-iQ — COMMENTED** [review]
With `clang-10` I can pass the magic bytes check (I guess it uses coverage in `memcmp` to solve for it?). Currently running the new policy_estimator_io fuzzer as well.

File: src/test/fuzz/util.h [hunk 3]
```diff
@@ -264,4 +265,122 @@ void InitializeFuzzingContext(const std::string& chain_name = CBaseChainParams::
...
+        }
+#ifdef _GNU_SOURCE
+        const cookie_io_functions_t io_hooks = {
+            FuzzedFileProvider::read,
+            FuzzedFileProvider::write,
+            FuzzedFileProvider::seek,
+            FuzzedFileProvider::close,
+        };
+        return fopencookie(this, mode.c_str(), io_hooks);
+#else
```
[the remark this answers is not in the document]
Comment (practicalswift) [reply src/test/fuzz/util.h:318]: This is intentional actually: I want exactly the same input bytes to be consumed by an `open()` call regardless of `fopencookie` being available or not.

File: src/test/fuzz/autofile.cpp [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+            WriteToStream(fuzzed_data_provider, auto_file);
+            break;
+        }
+        }
+    }
+    (void)auto_file.Get();
+    (void)auto_file.GetType();
+    (void)auto_file.GetVersion();
+    (void)auto_file.IsNull();
+    if (fuzzed_data_provider.ConsumeBool()) {
```
[the remark this answers is not in the document]
Comment (practicalswift) [reply src/test/fuzz/autofile.cpp:66]: Yes it is, but I want the fuzzer to also exercise the `release()` code path here, and it isn't called in the destructor :)

File: src/test/fuzz/buffered_file.cpp [hunk 5]
```diff
@@ -48,17 +49,30 @@ void test_one_input(const std::vector<uint8_t>& buffer)
                 break;
             }
             case 3: {
+                if (!opt_buffered_file->SetPos(fuzzed_data_provider.ConsumeIntegralInRange<uint64_t>(0, 4096))) {
+                    setpos_fail = true;
```
[the remark this answers is not in the document]
Comment (practicalswift) [reply src/test/fuzz/buffered_file.cpp:53]: Unfortunately I think a single failed `SetPos` call is sufficient to set a subsequent `FindByte` call into an infinite loop. A successful `SetPos` won't necessarily undo the harm of a prior failed `SetPos` call :)

File: src/test/fuzz/policy_estimator_io.cpp [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+}
+
+void test_one_input(const std::vector<uint8_t>& buffer)
+{
+    FuzzedDataProvider fuzzed_data_provider(buffer.data(), buffer.size());
+    FuzzedAutoFileProvider fuzzed_auto_file_provider = ConsumeAutoFile(fuzzed_data_provider);
+    CAutoFile fuzzed_auto_file = fuzzed_auto_file_provider.open();
+    // Re-using block_policy_estimator across runs to avoid costly creation of CBlockPolicyEstimator object.
+    static CBlockPolicyEstimator block_policy_estimator;
+    if (block_policy_estimator.Read(fuzzed_auto_file)) {
```
[the remark this answers is not in the document]
Comment (practicalswift) [reply src/test/fuzz/policy_estimator_io.cpp:25]: The thinking here is that a successful `Read` will "reset" the state of `block_policy_estimator`. Without the check we might be calling `Write(…)` on `block_policy_estimator` that is populated from a successful `Read` in a previous run: that would introduce unnecessary coverage instability.

More specifically this is the case I'm guarding against:

```
run n:
  block_policy_estimator.Read() returns true
  block_policy_estimator.Write()
run n+1:
  block_policy_estimator.Read() returns false
  block_policy_estimator.Write() # now with state from run n (instead of n+1)
```

File: src/test/fuzz/util.h [hunk 7]
```diff
@@ -264,4 +265,209 @@ void InitializeFuzzingContext(const std::string& chain_name = CBaseChainParams::
...
+
+    static int seek(void* cookie, int64_t* offset, int whence)
+    {
+        assert(whence == SEEK_SET || whence == SEEK_CUR); // SEEK_END not implemented yet.
+        FuzzedFileProvider* fuzzed_file = (FuzzedFileProvider*)cookie;
+        int64_t new_offset = 0;
+        if (whence == SEEK_SET) {
+            new_offset = *offset;
+        } else if (whence == SEEK_CUR) {
+            if (AdditionOverflow(new_offset, *offset)) {
```
[the remark this answers is not in the document]
Comment (practicalswift) [reply src/test/fuzz/util.h:361]: Oh, nice catch! Thanks for noticing: now fixed!

File: src/test/fuzz/policy_estimator_io.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (Crypt-iQ) [reply src/test/fuzz/policy_estimator_io.cpp:25]: Ok this makes much more sense. Otherwise it's not always possible to get the same coverage I think? Different machines on different runs may not hit the same `Read`->`Write`->`Read(false)` calls.

[the remark this answers is not in the document]
Comment (practicalswift) [reply src/test/fuzz/policy_estimator_io.cpp:25]: @Crypt-iQ Exactly! We want to have "coverage stability": a given input should result in the same coverage for a single run regardless of what has happened in previous runs. Achieving 100% coverage stability is not always possible. It might be due to architectural reasons (globals is a typical example), or it might be due to performance reasons (re-using an object might be required to get decent performance in terms of execs/s). We want to do what we can to at least minimise _unnecessary_ coverage instability.

**Comment by Crypt-iQ** [comment]
Can confirm that the new `policy_estimator_io` fuzzer is significantly faster with libfuzzer and hits checks in `Read` & `Write` that it otherwise doesn't hit in the `policy_estimator` fuzzer due to being slower. Also tried with AFL but was very slow since I forgot to compile with persistent mode :/. Will review the latest set of changes now. Coverage for just the `policy_estimator_io` fuzzer is here https://crypt-iq.github.io/policy_cov_llvm_outs/src/policy/fees.cpp.gcov.html

**Comment by Crypt-iQ** [comment]
Tested ACK ad6c348
